### ISSUE TITLE
Black picture with flimic rgb

### ISSUE THREAD

**Issue by cryptomilk** [darktable-org/darktable#3128]
**Describe the bug**

I've edited a few photos with 'filmic rgb' already. With the latest master 0ae2c5cc174df610c79bcfa1c5781f1f4f286192 those pictures are rendered to completely black. Disabling filmic rgb doesn't fix it. If I go in the history stack back to the item before filmic was enabled the first time it works again ...

I'm using OpenCL (AMD ROCm)

@aurelienpierre

**Comment by aurelienpierre** [comment]
OpenCL or C code path ?

**Comment by aurelienpierre** [comment]
What if you change the chroma preservation mode ?

**Comment by cryptomilk** [comment]
It stays black if I change the chroma preservation mode.

**Comment by cryptomilk** [comment]
Reverting 0d01bb0ef340a110b57f292eb12461a522257f48 doesn't fix it.

**Comment by aurelienpierre** [comment]
what if you move filmic before input profile or after output profile ? That could be related to RGB spaces conversions.

**Comment by cryptomilk** [comment]
Moving it before input profile works

**Comment by cryptomilk** [comment]
Moving it after output profile works too!

**Comment by cryptomilk** [comment]
Do you want the XMP?

**Comment by aurelienpierre** [comment]
So it's definitely linked to the colour conversion. 

If you put filmic back between input and output colour profile, but change the working RGB space (in input profile), does it change anything ? (please try internal RGB spaces like rec2020 and possibly spaces provided by ICC profiler in ~.config/darktable/colorin).

I have 2 bets here:
1. the colour profile from the RGB space is not loaded
1. AMD OpenCL doesn't have built-in dot multiplication support, so it doesn't like the change I have made.

**Comment by cryptomilk** [comment]
The input profile is the standard color matrix and the working profile is rec2020. Changing doesn't do anything to the image only to the histogram.

**Comment by TurboGit** [comment]
You may want to check again with master, an OpenCL fix has been put in place.

**Comment by flannelhead** [comment]
This wasn't fixed for me by https://github.com/darktable-org/darktable/pull/3129 unfortunately. I was also about to report this one.

**Comment by cryptomilk** [comment]
With master I get now:

```
/home/asn/workspace/projects/darktable/data/kernels/filmic.cl:231:1: warning: control may reach end of non-void function                                                           
}                                                                                        
^                                                                                        
1 warning generated.
```

**Comment by TurboGit** [comment]
@cryptomilk : this hasn't changed so I supposed you already had this warning before.

**Comment by TurboGit** [comment]
BTW, I cannot reproduce so maybe an iop-order issue. Can you reproduce this with new edits?

**Comment by cryptomilk** [comment]
data/kernels/filmic.cl:231 should call `return 0` for example

**Comment by aurelienpierre** [comment]
@cryptomilk what if you revert 26341e0 ?

**Comment by aurelienpierre** [comment]
@flannelhead what GPU do you run ?

**Comment by flannelhead** [comment]
I have Intel UHD 620, using the Intel Compute Runtime OpenCL driver on Arch Linux.

**Comment by cryptomilk** [comment]
@TurboGit 

```
data/kernels/filmic.cl:229:14: warning: double precision constant requires cl_khr_fp64, casting to single precision
      return 0.0;                                                                                                                                                                  
             ^                                                                                                                                                                     
1 warning generated.
```

Reverting 26341e0333eaef4769db00505eaf4d7995022760 doesn't fix it. I still get a black image ...

**Comment by cryptomilk** [comment]
@TurboGit it needs to be: `return 0.0f`

**Comment by TurboGit** [comment]
Fixed. Thanks.

**Comment by cryptomilk** [comment]
We can do a debugging session tomorrow if someone is interested ... :-)

**Comment by aurelienpierre** [comment]
I'm able to reproduce on Intel OpenCL, wich is weird because I developed with it.

**Comment by aurelienpierre** [comment]
Ok, issue don't affect colorbalance, so the changes done in `colorspace.cl` can be discarded from the possible.

Putting any native Lab module (tone curve, levels…) in the RGB parts of the pipe works properly, so the RGB -> Lab -> RGB path works.

Putting any native RGB module (filmic, channel mixer) in the Lab parts of the pipe breaks, so the Lab -> RGB -> Lab path is failing.

**Comment by cryptomilk** [comment]
Fun fact: the preview renders correctly :-)

**Comment by flannelhead** [comment]
@cryptomilk: do you mean the low-resolution view before the full-res image is rendered (pardon my ignorance, not yet familiar with darktable lingo)? If yes, that's the case for me as well. There are indeed interesting lines in darktable debug log if started with `darktable -d opencl`:
```
2,373503 [pixelpipe_process] [thumbnail] using device 0
3,933719 [pixelpipe_process] [full] using device 0
4,695126 [pixelpipe_process] [preview] using device -1
```
The preview seems to use a different device id.

**Comment by flannelhead** [comment]
Also something else that happens on my system: only with `chroma preservation = no`, the image is rendered black - in all the three chroma-preserving modes, it is light gray.

**Comment by aurelienpierre** [comment]
Previews usually use C SSE2 code path, unless you set the OpenCL perf profile to "very fast GPU".

I think we are dealing with low-level issues on memory allocation done differently between GPU vendors. I will keep investigating.

**Comment by cryptomilk** [comment]
I don't now if this needs to be a multiple of 2. But setting it to 2047, I get a black picture.

**Comment by parafin** [comment]
I think increasing opencl_memory_requirement can only have one effect - disabling OpenCL in darktable.

**Comment by cryptomilk** [comment]
@parafin can you explain in more detail? The GPU has 8GB of memory.

**Comment by cryptomilk** [comment]
@parafin Ok, you're right -> `0.297747 [opencl_init] discarding device 0 gfx803 due to insufficient global memory (8192MB).`

**Comment by aurelienpierre** [comment]
@cryptomilk could you checkout on a48ea3e8a8c9fd92b0073ff24b66be90f98cee18, enable the channel mixer module, ensure it is between input and output profile, and report if the image is still black ?

**Comment by cryptomilk** [comment]
@aurelienpierre Checking out a48ea3e, everything works just fine. The picture is rendered correctly, even if I enable the channel mixer.

**Comment by flannelhead** [comment]
Works for me since the merge of https://github.com/darktable-org/darktable/pull/3189 :)

**Comment by cryptomilk** [comment]
This has been fixed with #3189

**Comment by kanyck** [comment]
I still have it on 3.1.0+8~g7871a4f25-dirty.

FilmicRGB with color preservation set to RGB (norm) plus Local Contrast enabled caused black field instead of image. Disabling LC causes washed out colors (HDR-style looking), after few manipulation it gets to normal. Disabling OpenCL seems to fix it.

Steps to reproduce the behavior (darkroom mode):
1. Enable Filmic RGB
2. Set color preservation to RGB (norm)
3. Enable Local Contrast in default mode (laplacian)
4. Enjoy pure black

**Platform**
 - OS: [Gentoo linux]
 - Version GIT [3.1.0+8~g7871a4f25-dirty]
 - OpenCL activated 
 - Radeon RX470 Ellesmere + amdgpu + OpenCL blob from amdgpu-pro
